internal/builders/docker/main.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2022 SLSA Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+)
+
+func main() {
+	buildConfigPath := flag.String("build_config_path", "",
+		"Required - Path to a toml file containing the build configs.")
+	sourceRepo := flag.String("source_repo", "",
+		"Required - URL of the source repo.")
+	gitCommitHash := flag.String("git_commit_hash", "",
+		"Required - SHA1 Git commit digest of the revisoin for the soruce code to build the artefact from.")
+	builderImage := flag.String("builder_image", "",
+		"Required - URL indicating the Docker builder image, including a URI and image digest.")
+
+	flag.Parse()
+
+	// TODO(#1191): Do we need any verification steps?
+	// 1. sourceRepo is a valid URI
+	// 2. gitCommitHash is valid (we expect to see sha1: at the beginning of the input string)
+	// 3. Docker image is fully specified name@alg:digest
+	// 4. buildConfigPath is relative
+
+	artifacts := make(map[string]int)
+	artifacts["source"] = ArtifactReference {
+		URI: sourceRepo
+		Digest: toDigestMap(gitCommitHash)
+	}
+
+	imageParts := strings.Split(builderImage, "@")
+	if len(imageParts) != 2 {
+		log.Fprintln(os.Stderr, fmt.Errorf("got %s, want NAME@DIGEST format", builderImage))
+		os.Exit(1)
+	}
+
+	// TODO: check that imageParts[0] is a valid URI
+	artifacts["builderImage"] = ArtifactReference {
+		URI: imageParts[0]
+		Digest: toDigestMap(imageParts[1])
+	}
+
+	ep := ParameterCollection {
+		Artifacts: artifacts
+		Values: map[string]string{"configFile": buildConfigPath}
+	}
+
+	bd := BuildDefinition  {
```

Comment:
is the idea to create a `BuildDefinition`, and then execute it during the `BuildArtifact` command?

Reply:
Yes.

Reply:
SG -- I think creating a builder object with the BuildDefinition internally would work like below. Resolving!